content/chapters/io/lab/content/file-descriptors.md

```diff
@@ -0,0 +1,234 @@
+But what is a file descriptor exactly?
+From the application's point of view, it is a positive integer acting as a unique identifier for an I/O channel, such as a file.
+To the operating system, a file descriptor is an index.
+Each process has what's called a **file descriptor table**, which is just a fancy name for an array of pointers.
+A file descriptor is an index in this array.
+Each element in this array points towards a structure in the kernel's memory that represents an I/O resource.
+This structure is called the **open file strcuture**.
+
+To illustrate all this information, look at the image below.
+All data structures shown below are stored in the Kernel's memory area and are transparent from the user space.
+And remember that the file descriptor table is bound to the process, so all its threads share the same file descriptors.
+
+![File Descriptors](../media/file-descriptors.svg).
+
+To find out more about the contents of these structures, check out [this section in the Arena](./arena.md#open-file-structure-in-the-kernel)
+
+## Creating New File Descriptors
+
+We already know that each process gets 3 file descriptors "by default":
+
+- `stdin` (standard input): 0
+- `stdout` (standard output): 1
+- `stderr` (standard error): 2
+
+To create new file descriptors (i.e. open new files), a process to use the syscall [`open()`](https://man7.org/linux/man-pages/man2/open.2.html).
+It receives the path to the file, some flags which are akin to the mode string passed to `fopen()`, and a mode.
+This mode represents the permissions with which to create the opened file, it it doesn't exist.
+We'll get back to this mode in the future.
+
+From now, you should consult [`open()`'s man page](https://man7.org/linux/man-pages/man2/open.2.html) whenever you encounter a new argument to this syscall.
+Navigate to `support/file-descriptors/open_read_write.c`.
+The function `open_file_for_reading()` calls `open()` with the flag `O_RDONLY`, which is equivalent to opening the file with `fopen()` and the `"r"` mode.
+This means read-only.
+Note that the file descriptor we get is 3, just like before.
+
+Then `read_from_file()` reads `bytes_to_read` bytes from this file.
+For this, it uses the `read()` syscall.
+It returns the number of bytes that were read from the file.
+
+Notice that the code doesn't simply call `read(fd, buff, bytes_to_read)`.
+Instead, it uses a `while` loop to read data from the file.
+The reason is, as always, stated in [`read()`'s man page](https://man7.org/linux/man-pages/man2/read.2.html).
+The snippet below is about the number of bytes returned by `read()`:
+
+> It is not an error if this number is smaller than the number of
+> bytes requested; this may happen for example because fewer bytes
+> are actually available right now (maybe because we were close to
+> end-of-file, or because we are reading from a pipe, or from a
+> terminal), or because read() was interrupted by a signal.
```

Comment:
Makes sense. I'll keep this for the next part of the lab, about remote (network-based) I/O.

Reply:
I still strongly encourage the use of `while` loops and wrapper functions for I/O, so this information is really useful. But it may be worth adding examples both ways.